### ISSUE TITLE
Include the standard Qt translation files with our OSX installers

### ISSUE THREAD

**Issue by justinclift** [sqlitebrowser/sqlitebrowser#1626]
In #1592 we worked out how to add the missing standard Qt translations to our Windows installer.  We need to do the same for our OSX installers too.

**Comment by justinclift** [comment]
Looking into this now.  Adding the standard Qt files (trying with French for the moment) to the same `translations` directory as our other ones isn't working (on OSX).

Looks like they need to be in some other location.  Investigating...

**Comment by justinclift** [comment]
Having no joy at all with this.  Looking at how Qt Creator itself is deployed on OSX, it has a `Resources/translations/` directory, with all of the standard Qt translation files + its own in there.

Trying to copy that approach, if I copy the entire set of standard Qt translation files (from the Qt 5.11.2 translations dir) into a new `Resources/translations/` directory in our structure.  No change.

The way I'm **attempting** to test things is by:

1. Open the DB4S Preferences menu and change the language there (to French), then restart DB4S.
2. **Note** - The main DB4S interface is now in French (as expected)
3. Open any local database file, then click the (French equivalent) "New table" button.  This shows the "New table dialog" with the main strings in French.

Unfortunately, the "Cancel" and "OK" buttons at the bottom of the window are still in English:

![capture d ecran 2018-11-26 a 06 37 22](https://user-images.githubusercontent.com/406299/48996935-17238580-f1a2-11e8-8ef1-9c9d0d29f4c5.png)

I've also tried after changing the system language to French in the main OSX menu + restarting, so the entire OSX experience is in French.  No change, those "Cancel" and "OK" buttons are still in English for DB4S.

Also tried copying the entire set of Qt translation files into the DB4S `Contents/MacOS/translations/` folder.  Again no change at all.

Checking other applications while the OSX System setting is set to French, and they have correctly translated "Cancel" & "OK" buttons (using the French equivalent).

Anyone have ideas?  I'm hitting a blank wall atm. :frowning:

**Comment by justinclift** [comment]
Btw, thanks for the assist @thiagomacieira. :smile:

**Comment by justinclift** [comment]
The output from that is `/Applications/DB Browser for SQLite.app/Contents/translations`.

Thought I'd already tried that location, but I'll put the translations again now just in case and see what happens. :smile:

**Comment by justinclift** [comment]
Yep, that does look like the correct place after all.  I'd been trying slightly different path variations, missing that exact one.

Updating the build script to add the right files now...

**Comment by justinclift** [comment]
Initial manual testing of the new build script **seems ok**.  Running it end-to-end now.  If it completes ok, then new OSX nightlies will be online in ~2.5 hours.  They'll need testing (for each language) to make sure the language files are doing the right thing.  But, I'm hopeful it's mostly ok. :smile:

**Comment by justinclift** [comment]
The internet connection dropped out while those OSX nightlies were building, so that run failed.  Ran the build job again, this time successfully, and now I'm verifying the individual translations display ok when DB4S runs:

Language | Status
-- | --
Arabic | Unchecked
Chinese (China) | Unchecked
Chinese (Taiwan) | Unchecked
Czech | Unchecked
English (US) | Unchecked
English (UK) | Unchecked
French | Seems ok
German | Unchecked
Korean | Unchecked
Polish | Unchecked
Portuguese | Unchecked
Russian | Unchecked
Spanish | Unchecked
Turkish | Unchecked
Ukranian | Unchecked

**Comment by justinclift** [comment]
A should probably check each language for the Windows installers too, as the same problem is likely to happen there as well.

**Comment by justinclift** [comment]
Looking through the translation directory in the Qt 5.11.2 folder, it seems like the problem starts there.  There are just very few Chinese, Portuguese, and Turkish files.

I'm not (yet) sure if that's because of a build problem with the Qt 5.11.2 installers, or the Qt project just doesn't have translated strings for those languages yet.  Looking into it now. :smile:

**Comment by justinclift** [comment]
Yep, the problem seems to be the main Qt project not having translated strings for those languages:

&nbsp; &nbsp; https://github.com/qt/qttranslations/tree/5.11/translations

The longer term solution to the problem is probably to ask our translators for the missing bits, if they'd be ok to help translate for the main Qt project.  The results would reach us for the next Qt release.

As a cheating approach :wink: for now, we could possibly identify the strings needed by the DB4S dialogs and try to get translations done for just those.  Generating .qm files from those + adding them to our installers should fix things for us.

Naturally, those same strings would be useful for the main Qt project.

@bssthu @PeterDaveHello @firateski @bernardosulzbach Does any of the above sound like a reasonable idea? :smile:

**Comment by bernardosulzbach** [comment]
I'm OK with providing the patches for either the Qt project or to DB4S, I just need some better clarification about what you want me to do exactly. Also, I might have to solve some higher priority issues before I can work on this, but I will get to it eventually.

**Comment by justinclift** [comment]
Info on translating for the main Qt project itself seems to be here: https://wiki.qt.io/Qt_Localization

**Comment by justinclift** [comment]
Ugh.  It looks like contributing to the upstream Qt project could be a real pain:

&nbsp; &nbsp; https://lists.qt-project.org/pipermail/localization/2018-July/000459.html

They seem to need more than 1 person who can translate each language, as every change needs a reviewer in addition to the translator.  That seems like a blocker.

**Comment by justinclift** [comment]
This page, automatically updated daily, shows the .ts files for each Qt version:

&nbsp; &nbsp; http://l10n-files.qt.io/l10n-files/

It shows the percentage each translation file has been completed, and also provides templates for the untranslated files.

When loading the [untranslated Qt 5.11 "qtbase" file](http://l10n-files.qt.io/l10n-files/qt5-stable/qtbase_untranslated.ts) into Qt Linguist just now... it says there are 1600+ strings to be translated (wow!).  Qt Linguist has a search capability though, so it wasn't hard to find the likely correct "Cancel" and "OK" buttons in that. :smile:

![qt linguist finding specific strings we need](https://user-images.githubusercontent.com/406299/49056017-e731bc00-f24d-11e8-8dcb-66815aba4d9b.png)

If we can figure out the right strings to be translated, this shouldn't be too hard to get done. :smile:

We'll probably need to create a new git repo to hold this stuff, but that's pretty simple.

**Comment by justinclift** [comment]
Yeah, I think the number of strings we (DB4S) will need will be much, much shorter.  Probably much less than 100 for this bit. :smile:

**Comment by SafaAlfulaij** [comment]
Btw, the most recent Arabic translation in Qt was done by me, and I had to wait for some time till I can get a reviewer. At the end, there wasn't one, and they were fine +1 it :D

About the strings: sometimes strings from strange areas come to action. For example, the scrollbar context menu, edit lines and text lines context menus, errors connecting to network (in one of the recent builds), etc etc. Also many strings are just translations of keyboard keys, which might not take that much. Be selective and you'll finish fast.

**Comment by bernardosulzbach** [comment]
Thanks for your input, @SafaAlfulaij. However, I do think that @justinclift was suggesting that we translate internally instead of updating Qt, did I get it wrong?

**Comment by justinclift** [comment]
Yeah, that's the general line of thought I have.

eg first priority is to get the stuff done (for us) that we think we'll need.  We could also then provide the same strings to the Qt project, but we don't really need to care if they decide to reject them or something as it won't have a negative affect on us.

Time wise... I think it might be better to do this after the 3.11.0 release?  It could turn out to be complicated, and I'd rather not delay the .0 release for it. :smile:

**Comment by justinclift** [comment]
Excellent.  Sounds good to me too @bernardosulzbach. :smile:

**Comment by MKleusberg** [comment]
@justinclift Do you know if this is still an open issue? :smile:

**Comment by mgrojo** [comment]
According to the PR  #4005, the Qt translation is not being used on Windows. On Linux, I see that my own build on main branch works, but not our released 3.13.1 AppImage.

**Comment by lucydodo** [comment]
@mgrojo Was there an issue during the AppImage packaging process?

**Comment by PeterDaveHello** [comment]
I’m tied up with other commitments at the moment, so I haven’t had time to dig into the full Qt contribution process yet. If someone can outline a straightforward workflow for submitting translations upstream, or help open a dialog with the Qt maintainers about collaborating, I’ll definitely look at contributing the missing Traditional Chinese strings.

**Comment by mgrojo** [comment]
> [@mgrojo](https://github.com/mgrojo) Was there an issue during the AppImage packaging process?

I don't know. It's possible that the same underlying problem is present in both the AppImage and Windows packaging. Maybe we have to tell cmake to include the Qt translations files in the build?


> I’m tied up with other commitments at the moment, so I haven’t had time to dig into the full Qt contribution process yet. If someone can outline a straightforward workflow for submitting translations upstream, or help open a dialog with the Qt maintainers about collaborating, I’ll definitely look at contributing the missing Traditional Chinese strings.

These are two different issues. One is that we don't include the translation files already present in Qt, so we don't get any translation for any of the strings that are included in the Qt library, like default buttons and menus. Another matter is whether the Qt translation files are complete for each of the languages.

**Comment by lucydodo** [comment]
@mgrojo I'll look into this issue this weekend.

**Comment by lucydodo** [comment]
@mgrojo How did you install the Qt used during the build?
I still encounter the same issue when building from the master branch.

**Comment by mgrojo** [comment]
> [@mgrojo](https://github.com/mgrojo) How did you install the Qt used during the build? I still encounter the same issue when building from the master branch.

I just build it using the Ubuntu packages for Qt5, so I guess it takes the translation files available also on Ubuntu, I guess, from the qttranslations5-l10n package.

**Comment by ozgurerturk** [comment]
Hello, is this issue still open?

I encountered the same problem with turkish version of db browser for sqlite, I looked it up from qttranslations and found out they were added for the [5.15-tr branch](https://github.com/qt/qttranslations/tree/5.15-tr/translations). 

I generated the .qm turkish files, added them to my installed translations path and qt strings were translated correctly from my end. 

I guess it looks like Turkish and other catalogs are not included in qt 5.15.2 installation? Should I open a PR there or store the generated .qm files in this repo maybe?

**Comment by FriedrichFroebel** [comment]
> Hello, is this issue still open?

If the nightly builds still miss them, probably yes.

> Should I open a PR there or store the generated .qm files in this repo maybe?

This is most likely the wrong approach. Instead, they should be handled/included correctly on build time.
